File: java/Apparat/src/com/joa_ebert/apparat/tools/dump/Dump.java
/*
 * This file is part of Apparat.
 * 
 * Apparat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Apparat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with Apparat. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Copyright (C) 2009 Joa Ebert
 * http://www.joa-ebert.com/
 * 
 */

package com.joa_ebert.apparat.tools.dump;

import java.io.File;
import java.io.FileOutputStream;
import java.io.PrintWriter;

import com.joa_ebert.apparat.abc.Abc;
import com.joa_ebert.apparat.abc.utils.AbcPrinter;
import com.joa_ebert.apparat.swf.tags.ITag;
import com.joa_ebert.apparat.swf.tags.Tags;
import com.joa_ebert.apparat.swf.tags.control.DoABCTag;
import com.joa_ebert.apparat.swf.tags.define.DefineBitsJPEG2Tag;
import com.joa_ebert.apparat.tools.ITool;
import com.joa_ebert.apparat.tools.IToolConfiguration;
import com.joa_ebert.apparat.tools.ToolLog;
import com.joa_ebert.apparat.tools.ToolRunner;
import com.joa_ebert.apparat.tools.io.TagIO;

/**
 * 
 * @author Joa Ebert
 * 
 */
public final class Dump implements ITool
{
	public static void main( final String[] arguments )
	{
		final ToolRunner toolRunner = new ToolRunner( new Dump(), arguments );

		toolRunner.run();
	}

	private IToolConfiguration config;

	public String getHelp()
	{
		return "-input [file]\tThe input file\n"
				+ "-abc\tWill output detailed ABC information.\n"
				+ "-tags\t\tWill show known tags.\n"
				+ "-images\tWill extract DefineBitsJPEG2 images.";
	}

	public String getName()
	{
		return "Dump";
	}

	public boolean needsOutput()
	{
		return false;
	}

	public void run() throws Exception
	{
		if( config.getInput().endsWith( ".abc" ) )
		{
			final File input = new File( config.getInput() );

			final Abc abc = new Abc();

			abc.read( input );

			final File file = new File( input.getName() + ".txt" );

			ToolLog.info( "Exporting ABC to \"" + file.getName() + ".txt\"." );

			final FileOutputStream fileOutputStream = new FileOutputStream(
					file );

			new AbcPrinter( new PrintWriter( fileOutputStream ) ).print( abc );

			fileOutputStream.flush();
			fileOutputStream.close();
		}
		else
		{
			final TagIO tagIO = new TagIO( config.getInput() );

			tagIO.read();

			if( config.hasOption( "tags" ) )
			{
				ToolLog.info( "Tags:" );

				for( final ITag tag : tagIO.getTags() )
				{
					final String tagString = Tags.typeToString( tag.getType() );
					String output = tagString;

					if( tagString.length() < 5 )
					{
						output += "\t\t\t";
					}
					else if( tagString.length() < 12 )
					{
						output += "\t\t";
					}
					else
					{
						output += "\t";
					}

					ToolLog.success( output + tag.toString() );
				}
			}

			if( config.hasOption( "images" ) )
			{
				for( final ITag tag : tagIO.getTags() )
				{
					if( tag.getType() == Tags.DefineBitsJPEG2 )
					{
						final DefineBitsJPEG2Tag defineBitsJPEG2 = (DefineBitsJPEG2Tag)tag;

						String extension = "";

						if( defineBitsJPEG2.imageData[ 0 ] == (byte)0xff )
						{
							extension = ".jpg";
						}
						else if( defineBitsJPEG2.imageData[ 0 ] == (byte)0x89 )
						{
							extension = ".png";
						}
						else if( defineBitsJPEG2.imageData[ 0 ] == (byte)0x47 )
						{
							extension = ".gif";
						}

						new File( "images" ).mkdirs();

						final File file = new File( "images" + File.separator
								+ defineBitsJPEG2.characterId + extension );

						ToolLog.info( "Exporting JPEG image to \"images/"
								+ file.getName() + "\"." );

						final FileOutputStream fileOutputStream = new FileOutputStream(
								file );

						fileOutputStream.write( defineBitsJPEG2.imageData );
						fileOutputStream.flush();
						fileOutputStream.close();
					}
				}
			}

			if( config.hasOption( "abc" ) )
			{
				for( final ITag tag : tagIO.getTags() )
				{
					if( tag.getType() == Tags.DoABC )
					{
						final DoABCTag doABC = (DoABCTag)tag;
						final Abc abc = new Abc();

						abc.read( doABC );

						final File file = new File( doABC.name + ".txt" );

						ToolLog.info( "Exporting ABC to \"" + file.getName()
								+ "\"." );

						final FileOutputStream fileOutputStream = new FileOutputStream(
								file );

						new AbcPrinter( new PrintWriter( fileOutputStream ) )
								.print( abc );

						fileOutputStream.flush();
						fileOutputStream.close();
					}
				}
			}

			tagIO.close();
		}
	}

	public void setConfiguration( final IToolConfiguration configuration )
	{
		config = configuration;
	}
}

File: apparat-core/src/main/scala/apparat/tools/dump/Dump.scala
/*
 * This file is part of Apparat.
 *
 * Copyright (C) 2010 Joa Ebert
 * http://www.joa-ebert.com/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package apparat.tools.dump

import apparat.utils.IO._
import apparat.abc._
import apparat.abc.analysis.AbcUMLExport
import apparat.tools.{ApparatConfiguration, ApparatTool, ApparatApplication}
import java.io.{File => JFile, PrintWriter => JPrintWriter, FileOutputStream => JFileOutputStream}
import apparat.swf.{Swf, DoABC}
import apparat.bytecode.{BytecodeDump, BytecodeDumpTypeCFG, BytecodeDumpTypeDefault}

/**
 * @author Joa Ebert
 */
object Dump {
	def main(args: Array[String]): Unit = ApparatApplication(new DumpTool, args)

	class DumpTool extends ApparatTool {
		var input: JFile = _
		var output: Option[JFile] = None

		var exportSWF = false
		var exportUML = false
		var exportABC = false

		var bytecodeFormat: DumpBytecodeFormat = DumpBytecodeAsDefault

		var sourceName = ""

		override def name = "Dump"

		override def help = """	-i file				Input file (SWC, SWF or ABC)
	-o [dir]			Optional output directory
	-swf 					Basic SWF information
	-uml					UML diagrams in DOT format
	-abc					Detailed ABC information
	-bc [raw|cfg|default] Bytecode format (raw bytes, graphs, default)"""

		override def configure(config: ApparatConfiguration): Unit = configure(DumpConfigurationFactory fromConfiguration config)

		def configure(config: DumpConfiguration) = {
			input = config.input
			output = config.output
			exportSWF = config.exportSWF
			exportUML = config.exportUML
			exportABC = config.exportABC
			bytecodeFormat = config.bytecodeFormat
		}

		override def run() = {
			if(!exportSWF && !exportABC && !exportUML) {
				log.warning("Nothing specified to export. Try the -h option for help.")
			}

			val source = input

			output match {
				case Some(dir) => {
					dir.mkdirs
					assert(dir.isDirectory, "Given output path must point to a directory.")
				}
				case None =>
			}

			sourceName = source.getName

			if(sourceName.toLowerCase endsWith ".abc") {
				if(exportUML || exportABC) {
					exportAbcs((Abc fromFile source) :: Nil)
				}
			} else {
				val swf = Swf fromFile source

				if(exportSWF) {
					log.info("Writing SWF information ...")
					using(writerFor("info")) { swf dump _ }
				}

				if(exportABC || exportUML) {
					val abcs = swf.tags collect {
						case doABC: DoABC => Abc fromDoABC doABC
					}

					exportAbcs(abcs)
				}
			}
		}

		private def exportAbcs(abcs: List[Abc]) = {
			if(exportUML) {
				log.info("Writing UML graph ...")
				using(writerFor("uml")) { new AbcUMLExport(abcs) to _ }
			}

			if(exportABC) {
				bytecodeFormat match {
					case DumpBytecodeAsDefault => {
						abcs foreach { _.loadBytecode() }
						BytecodeDump.`type` = BytecodeDumpTypeDefault
					}
					case DumpBytecodeAsGraph => {
						abcs foreach { _.loadBytecode() }
						BytecodeDump.`type` = BytecodeDumpTypeCFG
					}
					case DumpBytecodeAsRaw =>
				}

				var i = 0

				for(abc <- abcs) {
					log.info("Writing ABC information ...")
					using(writerFor("abc"+i.toString)) { abc dump _ }
					i += 1
				}
			}
		}

		private def writerFor(name: String) = new JPrintWriter(new JFileOutputStream(outputFile(name)))

		private def outputFile(name: String) = output match {
			case Some(dir) => new JFile(dir, outputName(name))
			case None => new JFile(outputName(name))
		}

		private def outputName(name: String) = sourceName+"."+name+".txt"
	}
}
